### ISSUE TITLE
Unable to configure DestinationRules for routes from Gateway with 1.1.0-rc.0

### ISSUE THREAD

**Issue by robertpanzer** [istio/istio#12066]
**Describe the bug**
I created a service without the istio-proxy sidecar and wanted to make this available using the istio-ingressgateway.
To allows calls to this service I have to configure a DestinationRule that disables TLS for this connection as the service will not understand mTLS.
I configured the DestinationRule to use a dedicated subset name `plain`.

While the route is properly pushed to the ingress gateway pointing to `outbound|8080|plain|testserver.foo.svc.cluster.local`, the cluster doesn't appear there.

Additionally there is another cluster now that doesn't follow the naming scheme: `outbound_.8080_._.testserver.foo.svc.cluster.local`

I tried the same configuration with Istio 1.0.5 and that worked.

I can't get away from the feeling that I have done something wrong, but the fact that this config works with 1.0 and I can see no error made me create this Issue.

**Expected behavior**
It should be possible to expose services that don't have an istio-proxy via the istio-ingressgateway.

**Steps to reproduce the bug**
Install Istio 1.1.0-rc.0 and apply this configuration:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: foo
  labels:
    istio-injection: disabled

---
apiVersion: v1
kind: Namespace
metadata:
  name: foo2
  labels:
    istio-injection: enabled

---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: gateway-http
  namespace: foo
spec:
  selector:
    istio: ingressgateway
  servers:
    - port:
        number: 80
        name: http
        protocol: HTTP
      hosts:
        - test.server

---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: testserver
  namespace: foo
spec:
  gateways:
    - gateway-http
    - mesh
  hosts:
    - test.server
  http:
    - match:
        - uri:
            prefix: "/get"
      route:
        - destination:
            host: testserver.foo.svc.cluster.local
            subset: plain
            port:
              number: 8080
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: testserver
  namespace: foo
spec:
  host: testserver.foo.svc.cluster.local
  subsets:
    - name: plain
      trafficPolicy:
        connectionPool:
          tcp:
            connectTimeout: 2.000s
        tls: {}
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testserver
  namespace: foo
spec:
  selector:
    matchLabels:
      app: testserver
  replicas: 2
  template:
    metadata:
      labels:
        app: testserver
    spec:
      containers:
        - name: testserver
          image: kennethreitz/httpbin
          ports:
            - containerPort: 80
              name: http-testserver
---
apiVersion: v1
kind: Service
metadata:
  name: testserver
  namespace: foo
spec:
  selector:
    app: testserver
  ports:
    - name: http
      port: 8080
      protocol: TCP
      targetPort: http-testserver


---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: testcurl
  namespace: foo2
spec:
  selector:
    matchLabels:
      app: testcurl
  replicas: 1
  template:
    metadata:
      labels:
        app: testcurl
    spec:
      containers:
        - name: testcurl
          image: tutum/curl
          command:
            - sleep
            - "100000"
```

The service to expose is httpbin and resides in the namespace `foo` available as `testserver.foo.svc.cluster.local`.
The service exposes the application at port 8080.
A VirtualService for the host name `test.server` should route to this service.

After applying this configuration check the config of the ingress gateway.
The route properly routes to the correct cluster:
```json
   "dynamic_route_configs": [
    {
     "version_info": "2019-02-26T16:22:48Z/4",
     "route_config": {
      "name": "http.80",
      "virtual_hosts": [
       {
        "name": "test.server:80",
        "domains": [
         "test.server",
         "test.server:80"
        ],
        "routes": [
         {
          "match": {
           "prefix": "/get"
          },
          "route": {
           "cluster": "outbound|8080|plain|testserver.foo.svc.cluster.local",
```
However, this cluster is never pushed to the ingress gateway:
```
curl -s localhost:15000/clusters|cut -d ':' -f 1|uniq|sort|grep foo
outbound_.8080_._.testserver.foo.svc.cluster.local
outbound|8080||testserver.foo.svc.cluster.local
```

Hence I cannot call this service via the Ingress Gateway and get a 503 (which means it hits the VirtualService because it doesn't get a 404):
```
curl -v localhost/get -H 'host:test.server'
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 80 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 80 (#0)
> GET /get HTTP/1.1
> host:test.server
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 503 Service Unavailable
< date: Tue, 26 Feb 2019 16:35:09 GMT
< server: istio-envoy
< content-length: 0
<
* Connection #0 to host localhost left intact
```

The configuration above has a consumer deployment that has a sidecar.
As the VirtualService is also targeted to the mesh gateway, I can see there that the cluster is properly pushed to this sidecar:
```
curl -s localhost:15000/clusters|cut -d ':' -f 1|uniq|sort|grep foo
outbound|8080|plain|testserver.foo.svc.cluster.local
outbound|8080||testserver.foo.svc.cluster.local
```
Here I also don't see that strange looking cluster name with the `_.`.
Additionally I can call the VirtualService successfully from the consumer deployment:
```
root@testcurl-9ff7dd55d-w5z9t:/# curl -v 1.2.3.4/get -H 'host: test.server'
* Hostname was NOT found in DNS cache
*   Trying 1.2.3.4...
* Connected to 1.2.3.4 (1.2.3.4) port 80 (#0)
> GET /get HTTP/1.1
> User-Agent: curl/7.35.0
> Accept: */*
> host: test.server
>
< HTTP/1.1 200 OK
* Server envoy is not blacklisted
< server: envoy
< date: Tue, 26 Feb 2019 16:32:28 GMT
< content-type: application/json
< content-length: 807
< access-control-allow-origin: *
< access-control-allow-credentials: true
< x-envoy-upstream-service-time: 16
<
{
  "args": {},
  "headers": {
    "Accept": "*/*",
    "Content-Length": "0",
    "Host": "test.server",
    "User-Agent": "curl/7.35.0",
    "X-B3-Sampled": "1",
    "X-B3-Spanid": "211df12a8f8acac0",
    "X-B3-Traceid": "a8b1a1928c62b1e3211df12a8f8acac0",
    "X-Envoy-Decorator-Operation": "testserver.foo.svc.cluster.local:8080/get*",
    "X-Istio-Attributes": "CigKGGRlc3RpbmF0aW9uLnNlcnZpY2UubmFtZRIMEgp0ZXN0c2VydmVyCiYKHWRlc3RpbmF0aW9uLnNlcnZpY2UubmFtZXNwYWNlEgUSA2Zvbwo6Cgpzb3VyY2UudWlkEiwSKmt1YmVybmV0ZXM6Ly90ZXN0Y3VybC05ZmY3ZGQ1NWQtdzV6OXQuZm9vMgo8ChdkZXN0aW5hdGlvbi5zZXJ2aWNlLnVpZBIhEh9pc3RpbzovL2Zvby9zZXJ2aWNlcy90ZXN0c2VydmVyCj4KGGRlc3RpbmF0aW9uLnNlcnZpY2UuaG9zdBIiEiB0ZXN0c2VydmVyLmZvby5zdmMuY2x1c3Rlci5sb2NhbA=="
  },
  "origin": "10.1.4.207",
  "url": "http://test.server/get"
}
```

**Version**
```
k version
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:38:32Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:25:46Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
```
Istio 1.1.0-rc.0
The Ingress Gateway identifies like this:
```
"version": "fc273e117a6ed875c7cb9b3c3251c8d982bc7196/1.10.0-dev/Clean/RELEASE/BoringSSL"
```
The sidecar of the consumer deployment shows the same version

**Installation**
helm install ~/istio/istio-1.1.0-rc.0/install/kubernetes/helm/istio-init --name istio-init --namespace istio-system -f ~/istio/istio-1.1.0-rc.0/install/kubernetes/helm/istio/values-istio-auth-mcp.yaml  --set global.mtls.enabled=true --set global.tracing.enabled=true --set pilot.traceSampling=100.0 --set tracing.enabled=true --set grafana.enabled=true --set global.useMCP=true

helm install ~/istio/istio-1.1.0-rc.0/install/kubernetes/helm/istio --name istio --namespace istio-system -f ~/istio/istio-1.1.0-rc.0/install/kubernetes/helm/istio/values-istio-auth-mcp.yaml  --set global.mtls.enabled=true --set global.tracing.enabled=true --set pilot.traceSampling=100.0 --set tracing.enabled=true --set grafana.enabled=true --set global.useMCP=true

**Environment**
Running locally on Docker for Mac.

**Cluster state**
The attached file contains the configuration of the Ingress Gateway on 1.1 and 1.0.
Additionally it contains the logs of the Pilot on 1.1.0-rc.0.
[logs.zip](https://github.com/istio/istio/files/2906516/logs.zip)

**Comment by yangminzhu** [comment]
Seems due to #12000 . A quick workaround is to check if you have a default destination rule in istio-system (you should as you `--set global.mtls.enabled=true`), and you can create some default sidecar resource with the following contents:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: default
  namespace: foo
spec:
  egress:
  - hosts:
    - "foo/*"
    - "foo2/*"
    - "istio-system/*"
```

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: Sidecar
metadata:
  name: default
  namespace: foo2
spec:
  egress:
  - hosts:
    - "foo/*"
    - "foo2/*"
    - "istio-system/*"
```

**Comment by robertpanzer** [comment]
I cannot confirm that this works.
I still don't see my plain cluster but the two I mentioned above.

**Comment by costinm** [comment]
Sidecar should fix the problem - but should not be required.

Let's keep this bug for the case of 'no sidecar used'.

**Comment by robertpanzer** [comment]
Maybe it should, but it didn't for me, maybe I did something wrong.
I also tried to add a Sidecar resource that selects the Ingress Gateway, because my understanding is that I configure the Sidecar to select the affected Workload, here the Ingress Gateway, and tell it to "collect" routing rules from other namespaces.

**Comment by howardjohn** [comment]
I was also unable to get it working, even with the Sidecar. I'll looking into it some more today to see what is wrong, but it looks like it may be a different issue than #12000

**Comment by howardjohn** [comment]
To simplify this, I created a Deployment+Service+VirtualService+Gateway+DestinationRule as shown here for testserver. With injection enabled, I can access it through the ingressgateway fine. When run the server without a sidecar, requests fail with a 503 and `upstream connect error or disconnect/reset before headers`. Adding a Sidecar resource does not fix this.

Swapping out the httpbin server for an image with tcpdump shows that the pod is receiving traffic. It seems like it is responding with 
```
HTTP/1.1 400 Bad Request
Content-Type: text/plain; charset=utf-8
Connection: close

400 Bad Request
```
and this is sent back as a 503 from Envoy.

**Comment by duderino** [comment]
@philrud will take a crack at this

**Comment by philrud** [comment]
Here's what I found so far: the DestinationRule works if it's moved to the namespace where ingress gateway resides. This is likely a regression introduced by "exported destination rules" functionality (first part of which was introduced in 1.1). I'm investigating if there's a low-risk fix for this that could still make it into 1.1.

**Comment by robertpanzer** [comment]
:+1:
Do you think this only affects the ingress gateway?
Or does that mean that I have to also create the DestinationRule in all other namespaces where another „mesh“ gateway consumer lives.

**Comment by duderino** [comment]
@philrud nice progress

**Comment by philrud** [comment]
@robertpanzer this will only affect proxies/services that reside in istio-system (or whatever namespace was used for Istio components during installation), so assuming typical installation this will affect the gateway, but not the services within the mesh (considering they're outside of istio-system ns).

The cause is the [changeset from around a month ago](https://github.com/istio/istio/commit/09aa7c32b14a49b9688271dcd4eea90ddb2b8cf3#diff-ee39cd332931ab6f8c10713ce3446d09R535) that introduced quite a serious change of the behavior (regression?) in the DestinationRule-s resolution logic (Sidecar-less case). Basically now DestinationRule-s residing in the same namespace as the proxy will trump all other rules from other namespaces, even if those rules are more specific.

The problem is that the default Istio installation with gateways and mTLS=enabled will put the gateway and the default global mTLS DestinationRule into the same namespace (i.e. istio-system), meaning there will be no way to override that wildcard rule for gateway with a more specific one as was possible before (unless the more specific one also resides in the istio-system ns). Basically the current Istio.io examples recommend dropping a DR into the service namespace in order to turn off mTLS, so I'm sure somebody will get burned by this.

@rshriram could you please provide some background? I assume change of the behavior was intentional, but I can't tell for sure as tests are missing/commented out. 

One compromise I'm thinking of – maybe we can partially restore the behavior of choosing more specific rule, but only consider the proxy and service namespaces? So we can override the wildcard mTLS rule by a rule in the service namespace, but not an arbitrary namespace? This will be a pretty small change, [here's a draft](https://github.com/philrud/istio/commit/ac8040f8486c6ddf611d3b8239c19c6f0caed699). 

Do we actually want to get this fixed in 1.1 or we can get away with recommending users to move the DR's to 'istio-system'? There may be other implications of such change of course. @rshriram @costinm @duderino thoughts?

**Comment by howardjohn** [comment]
@philrud Before we mentioned adding a Sidecar config would workaround this. Based on what you said I think it still should if it's in the Istio namespace. Not sure without testing, the code is somewhat complex. I am working on a similar issue, we should add some testing so we don't run into these issues anymore

**Comment by rshriram** [comment]
@philrud your assessment is spot on.. and the change set was intentional - the prior mode of destination rule selection was very arbitrary. #12216 has a stop gap fix but it is likely to blow up else where. 

cc @louiscryan FYI..
The proper fix is to set the istio-config as the config root namespace by default, and move things like *.local DR rules into this namespace.

Now that said, @robertpanzer, as an end user, does the selection mechanism below sound reasonable?
1. pick best matching dest rule from the calling sidecar/gateway’s own namespace
2. if there is no (1), pick best matching dest rule from the target service’s namespace
3. pick best matching dest rule from global config root namespace (e.g., `istio-config`).

**Comment by robertpanzer** [comment]
@rshriram Let me see if I understood this correctly:

If I assume that mTLS is enabled, that means there is a default DestinationRule in the namespace Istio-system.
The IngressGateway is also usually in the same namespace.

Does that mean that every DestinationRule to any service from the Ingress Gateway must be configured in istio-system?
If this is this case this doesn't seem right to me.
In my opinion write access to istio-system should be restricted as much as possible.

**Comment by duderino** [comment]
@rshriram @philrud is there anything left to do here?

**Comment by philrud** [comment]
@duderino I have verified @rshriram 's fix -- it does address the specific issue that @robertpanzer has reported, so as long as @costinm and @louiscryan are fine with this for 1.1 we can close this issue

**Comment by duderino** [comment]
@philrud @rshriram thanks!

**Comment by louiscryan** [comment]
@robertpanzer to address your question above

1. pick best matching dest rule from the calling sidecar/gateway’s own namespace **(unless this calling namespace is also the global config root namespace)**
2. if there is no (1), pick best matching dest rule from the target service’s namespace
3. pick best matching dest rule from global config root namespace (e.g., istio-config).

For the default ingress gateway installed into istio-system we hit the corner case I bolded above and so the rule resolution falls into step no. 2 and would match the DestinationRule from the namespace where the referenced service was defined.

As a general rule we expect resolution to typically occur using step 2 or 3 and that an override of the DestRule by the consuming namespace is less common.

by example...


Namespace: istio-system (global config root)
  Gateway (name=ingress, host=*)
  VirtualService(host=*.example.com, gateway: ingress, route '/' -> example.sales.svc.cluster.local)
  DestRule(host=*, mTLS=off, LB=round-robin)

Namespace: something-internal
  Sidecar(egress : example.sales.svc.cluster.local)
  DestRule(host=*, mTLS=off, LB=random)

Namespace: sales
 Service(example.sales.svc.cluster.local)
 DestRule(host=*, mTLS=on, LB=round-robin)



The DestRule applied to the 'ingress' Gateway will be the one defined in 'sales' (istio-system is the config root) and so we resolved from the namespace where examples.sales.svc.cluster.local is defined.

The DestRule applied to the Sidecar in namespace 'something-internal' will be the one from that namespace (its not the config root namespace) as the host name rule matches 'examples.sales.svc.cluster.local'. If this was not the intended effect the owner of the namespace could correct it by refining the host matching rule to something more selective, e.g. *.something-internal.svc.cluster.local that only selects the host names they want to apply the rule to

Overall this setup achieves a resolution hierarchy of

service consuming namespace -> service defining namespace -> admin config namespace

with a special carveout when admin namespace == service consuming namespace to avoid unintended consequences because we have co-mingled implementation and global config. As a general rule we will recommend that teams move away from using 'istio-system' as their global configuration root. Expect to see further guidance on this going forward.
